### PR TITLE
Add experimental support for snapshot/restore

### PR DESCRIPTION
**Pull request by anuraaga**
This allows a user to opt-in to having snapshot/restore functionality exposed to host functions. This can be used to implement host-based exception handling, for example with `wasix`'s implementation of `setjmp` and `longjmp`. The basic machinery can also likely be reused if implementing WebAssembly exceptions in the future.

This is intended to stay in experimental as there is no official standard for exceptions yet, but as there are compilers such as wasix SDK that can compile them, this unblocks allowing wazero to run such programs. For example, the sqlc tool is popular in Go but has cgo issues, this is a step towards using wasix to compile it's Postgres parser and have a pure Go sqlc binary.

Discussed in this slack thread: https://gophers.slack.com/archives/C04CG4A2NKX/p1697555612115509?thread_ts=1697441543.851059&cid=C04CG4A2NKX

wasix
stack_checkpoint: https://wasix.org/docs/api-reference/wasix/stack_checkpoint
stack_restore: https://wasix.org/docs/api-reference/wasix/stack_restore

Example usage in wasix built app:
https://github.com/wasilibs/go-pgquery/blob/main/internal/wasix_32v1/wasix.go#L130

### CODE REVIEW

File: experimental/checkpoint_test.go
Comment (anuraaga) [inline experimental/checkpoint_test.go:120]: Assuming `ctx` is propagated through a nested invocation, I think it'd be possible to handle this as an error, but I don't know if it's worth it since it would be an error either way but add some tricky handling

**Comment by ncruces** [comment]
Issue #1495 is relevant.

File: internal/engine/compiler/engine.go
Comment (mathetake) [inline internal/engine/compiler/engine.go:1078]: this defer causes allocation every time a Go host func is called. would it be possible to avoid when snapshot is disabled?

**Comment by anuraaga** [comment]
@ncruces Just to refresh memory :)

https://gophers.slack.com/archives/C04CG4A2NKX/p1698068979958989?thread_ts=1697441543.851059&cid=C04CG4A2NKX

The biggest reason for holding off at the time was that a "real world test", aka using wasix-sdk, requires enabling threads. Now that wazero has threads support, the issue is gone or at least reduced.

File: internal/engine/compiler/engine.go
Comment (ncruces) [inline internal/engine/compiler/engine.go:1084]: Can we do a single `ok && s.ce == ce` and merge the else clauses?

**Comment by anuraaga** [comment]
Great point, that is indeed my intention so I should verify with a replace directive, I'll check it tomorrow.

File: experimental/checkpoint.go
Comment (anuraaga) [reply experimental/checkpoint.go:21]: I followed the pattern of function listener when writing this, I could change it to the sock pattern instead though. Any suggestion on what package the internal context key would go in?

File: internal/engine/compiler/engine.go
Comment (anuraaga) [reply internal/engine/compiler/engine.go]: Sweet, thanks!

**Review by anuraaga — COMMENTED** [review]
@ncruces Here I confirm replacing wazerox with wazero

https://github.com/wasilibs/go-pgquery/pull/23

As for the juggling, I guess it's mostly just stubbing out unused functions that make it into the binary

https://github.com/wasilibs/go-pgquery/blob/main/internal/wasix_32v1/wasix.go

I don't think we can say it's really easy to use wasix with wazero but I guess it's not terrible either

**Comment by ncruces** [comment]
> I don't think we can say it's really easy to use wasix with wazero but I guess it's not terrible either

This is good enough. I mean, we can't correctly extend wasi into wasix outside wazero because em we'd need access to internals of wasi implementation (e.g. filesystem).

But you can usefully stub stuff emitted by wasix-sdk, or you can likely implement wasix from scratch outside wazero (same as stealthrocket/wasi-go).

Good enough for me.

**Review by ncruces — APPROVED** [review]
I'd like a second opinion on the enabling API, but other than that, LGTM.

File: experimental/checkpoint_test.go
Comment (mathetake) [inline experimental/checkpoint_test.go:13]: could you move this to internal/integration_test/adhoc.go? That's where we ensure the consistency across engines.

File: internal/engine/compiler/engine.go
Comment (mathetake) [inline internal/engine/compiler/engine.go:794]:
```suggestion
	snapshotEnabled := ctx.Value(experimental.EnableSnapshotterKey{}) != nil
	if snapshotEnabled {
		ctx = context.WithValue(ctx, experimental.SnapshotterKey{}, ce)
	}
```

File: experimental/checkpoint_test.go
Comment (anuraaga) [reply experimental/checkpoint_test.go:13]: Sure I can try it. Two issues

- This reuses the same snapshot.wasm used for the example test. Should I just copy the wasm?
- It's highly experimental, is it ok to have tests in the core area?

File: internal/engine/compiler/engine.go
Comment (anuraaga) [reply internal/engine/compiler/engine.go]: 🤦

File: experimental/checkpoint_test.go
Comment (mathetake) [reply experimental/checkpoint_test.go:13]: ok let's leave it as-is now!

**Review by mathetake — APPROVED** [review]

**Pull request merged**
